File: JULY-2025/06-07-2025/p1865.js
var FindSumPairs = function(nums1, nums2) {
    this.nums1 = nums1;
    this.nums2 = nums2;
    this.freqMap = new Map();

    for (let num of nums2) {
        this.freqMap.set(num, (this.freqMap.get(num) || 0) + 1);
    }
};

FindSumPairs.prototype.add = function(index, val) {
    const oldVal = this.nums2[index];
    const newVal = oldVal + val;

    // Update frequency map
    this.freqMap.set(oldVal, this.freqMap.get(oldVal) - 1);
    if (this.freqMap.get(oldVal) === 0) {
        this.freqMap.delete(oldVal);
    }

    this.nums2[index] = newVal;
    this.freqMap.set(newVal, (this.freqMap.get(newVal) || 0) + 1);
};

FindSumPairs.prototype.count = function(tot) {
    let count = 0;
    for (let num of this.nums1) {
        let complement = tot - num;
        if (this.freqMap.has(complement)) {
            count += this.freqMap.get(complement);
        }
    }
    return count;
};

File: JULY-2025/06-07-2025/p1865.scala
class FindSumPairs(_nums1: Array[Int], _nums2: Array[Int]) {
  import scala.collection.mutable

  private val nums2 = new mutable.HashMap[Int, Int](10_000, mutable.HashMap.defaultLoadFactor)

  {
    _nums2.foreach(n => patch(n, 1))
  }

  private def patch(num: Int, count: Int): Unit =
    nums2.updateWith(num) {
      case Some(prev) => Some(prev + count)
      case None       => if (count < 0) None else Some(count)
    }

  def add(index: Int, `val`: Int): Unit = {
    val prevNum = _nums2(index)
    val newNum = prevNum + `val`
    
    _nums2(index) = newNum

    patch(prevNum, -1)
    patch(newNum, 1)
  }

  def count(tot: Int): Int =
    _nums1.foldLeft(0){ case (acc, n) =>
      acc + nums2.getOrElse(tot - n, 0)
    }

}